mailbox/noise.go

```diff
@@ -690,29 +711,62 @@ func (b *Machine) GenActTwo() ([ActTwoSize]byte, error) {
 	//
 	// However, we also want to send the client data they may need for
 	// authentication (if present) encrypted with strong forward secrecy.
-	var payload [ActTwoPayloadSize]byte
-	if b.authData != nil {
-		// If we have an auth payload, then we'll write out 2 bytes
-		// that denotes the true length of the payload, followed by the
-		// payload itself.
-		var payloadWriter bytes.Buffer
+	var (
+		actTwo      []byte
+		authPayload []byte
+	)
+	if b.handshakeVersion == HandshakeVersion0 {
+		actTwo = make([]byte, ActTwoSizeV0)
+		var payload [ActTwoPayloadSizeV0]byte
+		if b.authData != nil {
+			// If we have an auth payload, then we'll write out 2
+			// bytes that denotes the true length of the payload,
+			// followed by the payload itself.
+			var payloadWriter bytes.Buffer
+
+			var length [authDataLenSizeV0]byte
+			binary.BigEndian.PutUint16(
+				length[:], uint16(len(b.authData)),
+			)
+
+			_, err := payloadWriter.Write(length[:])
+			if err != nil {
+				return nil, err
+			}
+			_, err = payloadWriter.Write(b.authData)
+			if err != nil {
+				return nil, err
+			}
+
+			copy(payload[:], payloadWriter.Bytes())
+		}
+
+		authPayload = b.EncryptAndHash(payload[:])
+
+	} else {
```

Comment:
Should be a `switch` statement. Assumes the only other version will be 1 atm. On default we should bail out.